### ISSUE TITLE
Empty U-SRV database

### ISSUE THREAD

**Issue by bergeronian** [joshp23/YOURLS-IQRCodes#44]
Installation succeeds and shows a good installation of U-SRV in the YOURLS IQRPlugin Config, but the database is blank and QR codes do not appear.

**Comment by bergeronian** [comment]
YOURLS 1.8, U-SRV 2.3.3, IQRPlugin 2.2.1.

If I use the mass QR check, it will create the cache files, but not update the database.  I can't retrieve a QR code using the .qr URL method either.
If I try to create a new link, it does not create a cached image or update the database.  
If I use the U-SRV plugin to upload a test file, it writes the cache and updates the database.

**Comment by joshp23** [comment]
> the database is blank and QR codes do not appear.

The qr codes are not put in the U-SRV database, they are stored as files in the cache folder so that much is ok. When you say that they don't appear here, what do you mean, in the file system, in the YOURLS ui, or both?
> the mass QR check, it will create the cache files  

After you mass check, do the qr codes appear in YOURLS share boxes?

> If I try to create a new link, it does not create a cached image

Any log data, or web console output?

**Comment by bergeronian** [comment]
My apologies Josh,

It looks like now, after reverting to 1.79 and then 1.80, things have changed but I still have the same issue within the UI.  If I create a new link, the qr IS now generated in the cache, I can go to the server and view it.  The QR does not show in the UI when I create a new link.

[cid:132ff3a7-a14f-40bb-a2d6-531d83cbaf93]

Trying to access the QR with https://lnk.mcla.edu/nco49.qr results in a broken image.  Maybe my logic is bad here, but it sounds more like I can't retrieve the image from U-SRV.

I will look for more logs but I have nothing when I turn on debug in YOURLS, scanning the access.log or the php_error.log.

Any thoughts you have would be appreciated.

Thank you,

________________________________

**Comment by bergeronian** [comment]
> After you mass check, do the qr codes appear in YOURLS share boxes?

No, they do not.  Broken link. I think this has more to do with USRV retrieving files.  When I attempt to download a file by generating my own key, I get a file but it is empty in the case of a test file made from within the USRV plugin, and a corrupt image in the case of a QR.  The access.log shows a successful attempt, result code 200.

**Comment by joshp23** [comment]
1. If you are using NGINX, please make sure that you follow [this](https://github.com/joshp23/YOURLS-IQRCodes#note) instruction. If you are using Apache and SSL connections, try [this](https://github.com/joshp23/YOURLS-U-SRV#use-notes) instruction. Did you try this, did it fix the issue?
2. Just to be clear, you can confirm the following is working
    - QR code generation when adding a new link
    Yes?
    - QR code generation when using `mass check` function   
    Yes?
3. You have neither been able to cause a QR Code to display in the UI where the Share Box appears, nor by using the `.qr` method. No errors are showing in your access.log or error.log. 
  Yes?
4. Somewhere there is a failed attempt to fetch an image. You might be able to trace this with the Web Developer Console in FIrefox or Chrome. 
  a. Open up your access/error logs for YOURLS. Turn on YOURLS `debug` and all relevant error reporting.
  b. Open a fresh tab (clear cache, etc) and open the Web Dev Console for Firefox or Chrome.
  c. Attempt to access `https://lnk.mcla.edu/nco49.qr` and be mindful of the Web Dev Console output.
  d. Refresh your log files and read through the data there and the console.

In the Web Developer Console there should be  2 or 3 get requests. The first, once expanded, should show a `301 Moved Permanently` status. Is that showing for you? I should change this to 302, but it is what it is for now. Anyway, the second should be the usrv link. That's what we care about. You should be getting some kind of error there. Here's an example of the output I'm looking for:
```
GEThttps://example.com/srv/?id=iqrcodes&key=4fca69185aec7eb983eca347f23b6a99&fn=qrc_d716aha879a3adb408di6a6873afde5d.png
[HTTP/2 200 OK 150ms]

	
GET
	
scheme
	https
host
	example.com
filename
	/srv/
id
	iqrcodes
key
	4fca69185aec7eb983eca347f23b6a99
fn
	qrc_d716aha879a3adb408di6a6873afde5d.png
Address
	[XXXX:XXXX:XXXX:XXXX::XXXX]:443
Status200
OK
VersionHTTP/2
Transferred2.14 KB (1.62 KB size)
Referrer Policyno-referrer
```

**Comment by bergeronian** [comment]
Thank you for sticking with me on this.  I was getting a 200 on image load, FWIW.  What I found eventually, was that the file would download successful, with two line feeds at the beginning.  I couldn't find anything in srv.php that would introduce blank lines but I "solved" it by adding:

```
ob_clean();
flush();  
```

On line 159 of srv.php, just before the readfile($file).

**Comment by joshp23** [comment]
> Thank you for sticking with me on this. 

Indeed.
> I was getting a 200 on image load, FWIW.
Ok. I was wanting to explicitly rule out IQRCodes errors, and zero in on any subtle networking issues, or rule that out. So that U-SRV could be singled out, and you beat me to it.
```
ob_clean();
flush();  
```
seems like a pull request in the making. ;)
